### PR TITLE
Add multinode support to benchmark federates using a common base federate

### PR DESCRIPTION
**Pull request by nightlark**
### Summary
If merged this pull request will convert the various benchmark federates to a shared base class that provides common functionality, and enables easier reuse in a multinode benchmark scenario.

### Proposed changes
- Add a BenchmarkFederate base class for the benchmarks
- Convert existing benchmarks to use the BenchmarkFederate as a base class
- Add standalone executables for the benchmark federates, for use in multinode benchmark runs
- Print the number of CPUs on the machine

### CODE REVIEW

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Clones added
============
- benchmarks/helics/echoMessageBenchmarks.cpp  1
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -15
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
         

Clones removed
==============
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/EchoHubFederateMain.cpp  2
         

Clones removed
==============
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/ringBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
         

Clones removed
==============
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/ringBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
         

Clones removed
==============
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

File: benchmarks/helics/BenchmarkFederate.hpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
+/*
+Copyright (c) 2017-2020,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC.  See
+the top-level NOTICE for additional details. All rights reserved.
+SPDX-License-Identifier: BSD-3-Clause
+*/
+
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkFederate.hpp]: Is there are reason we aren't using #pragma once here,  we have been moving away from include guards?

File: benchmarks/helics/BenchmarkFederate.hpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    void setMaxIndex(int i) { maxIndex = i; }
+    /** gets the max index parameter*/
+    int getMaxIndex() { return maxIndex; }
+
+    // functions for setting callbacks
+    /** sets a callback function to call immediately after doMainLoop() returns, but before
+     * the helics finalize() call
+     * @param cb a function that takes no arguments and returns void
+     */
+    void setBeforeFinalizeCallback(std::function<void()> cb = {}) { callBeforeFinalize = cb; }
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkFederate.hpp]: given you are passing by value this could probably be a move into callBeforeFinalize.

File: benchmarks/helics/BenchmarkFederate.hpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+  protected:
+    class Result {
+      public:
+        std::string name;
+        std::string key;
+        std::string value;
+    };
+    std::vector<Result> results; //!< a vector of output results to print
+
+    OutputFormat result_format = OutputFormat::PLAIN_TEXT; //!< output format for printing results
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkFederate.hpp]: I have been trying to move more toward brace initialization for value in classes, so for at least new stuff we should use {...}  instead of =.. for initializers.

File: benchmarks/helics/BenchmarkFederate.hpp [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+    // parameters most benchmark federates need
+    helics::Time deltaTime = helics::Time(10, time_units::ns); //<! sampling rate
+    helics::Time finalTime = helics::Time(10000, time_units::ns); //<! final time
+    int index = 0; //<! the index for an instance of the benchmark federate
+    int maxIndex = 0; //<! the maximum index + 1 given to a benchmark federate in a run
+
+    // CLI11 Options for derived classes to change them if needed (e.g. set required)
+    CLI::Option* opt_delta_time; //<! the CLI11 option for --delta_time
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkFederate.hpp]: these should be initialized to nullptr since I believe we have a default constructor

File: benchmarks/helics/EchoHubFederate.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#include "helics/application_api/Publications.hpp"
+#include "helics/application_api/Subscriptions.hpp"
+
+using helics::operator"" _t;
+/** class implementing the hub for an echo test*/
+class EchoHub: public BenchmarkFederate {
+  private:
+    std::vector<helics::Publication> pubs;
+    std::vector<helics::Input> subs;
+    int cnt = 10;
```
Comment (phlptp) [inline benchmarks/helics/EchoHubFederate.hpp]: this should be named something more descriptive and brace initialized with the default

File: benchmarks/helics/EchoHubFederate.hpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  public:
+    EchoHub(): BenchmarkFederate("echo hub benchmark federate") {}
+
+    void setupArgumentParsing() override
+    {
+        // default final time for this benchmark
+        finalTime = helics::Time(100, time_units::ms);
+
+        app->add_option("--num_leafs", cnt, "the number of echoleaf federates to expect")
```
Comment (phlptp) [inline benchmarks/helics/EchoHubFederate.hpp]: I would add a true to the call, so the 10 (default) shows up.  
Why is this required?

File: benchmarks/helics/EchoLeafFederateMain.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    fed.setBeforeFinalizeCallback([&end_time]() { end_time = std::chrono::steady_clock::now(); });
+
+    // run the benchmark
+    fed.run([&start_time]() { start_time = std::chrono::steady_clock::now(); });
+
+    // print duration
+    auto elapsed =
+        std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time).count();
+    fed.addResult<decltype(elapsed)>("ELAPSED TIME (ns)", "real_time", elapsed);
+    fed.printResults();
```
Comment (phlptp) [inline benchmarks/helics/EchoLeafFederateMain.cpp]: Is there going to be any difference in the output from the hub to leafs logs so we know which is which?

**Comment by phlptp** [comment]
I will work my way through this over the weekend

File: benchmarks/helics/EchoMessageHubFederateMain.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (phlptp) [inline benchmarks/helics/EchoMessageHubFederateMain.cpp]: Would there be any sense in just having a single main with an argument which one to run, since a lot of these main operations look pretty similar now and have a common base class?

File: benchmarks/helics/PholdFederate.hpp [hunk 9]
```diff
@@ -39,60 +34,58 @@ class PholdFederate {
...
-    // callbacks for instrumenting the code
-    std::function<void()> callBeforeFinalize = nullptr;
-    std::function<void()> callAfterFinalize = nullptr;
-
-    bool initialized{false};
-    bool readyToRun{false};
-
   public:
-    PholdFederate() = default;
+    // TODO: output cluster name in a summary file, along with node count and feds per node
```
Comment (phlptp) [inline benchmarks/helics/PholdFederate.hpp]: does this TODO need to be done in this PR or is that something that will wait a while?

File: benchmarks/helics/testVirtVar.cpp [hunk 10]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (phlptp) [inline benchmarks/helics/testVirtVar.cpp]: What is this for,  looks like a bit of test case that got included?

**Comment by phlptp** [comment]
Overall I really this simplification.  I wonder if it makes sense to merge all the benchmarks to a single executable with a subcommand argument and a benchmark factory.  Then we don't have a whole bunch of very similar looking main functions around if we need to change something or add another benchmark.

File: benchmarks/helics/EchoHubFederate.hpp (hunk 6 above)
Comment (nightlark) [reply benchmarks/helics/EchoHubFederate.hpp]: I marked the parameters that were previously passed in using the `initialize` calls as required. This one probably doesn't need to be required since the default is okay.

File: benchmarks/helics/EchoLeafFederateMain.cpp (hunk 7 above)
Comment (nightlark) [reply benchmarks/helics/EchoLeafFederateMain.cpp]: There should be a difference -- do you think a default result entry called ("Federate Name") should be added to distinguish them, or should the launcher name the files with the federate type?

The file naming should probably be done either way, so it's mostly a question of whether the federate should print the name as part of the file contents.

File: benchmarks/helics/PholdFederate.hpp (hunk 9 above)
Comment (nightlark) [reply benchmarks/helics/PholdFederate.hpp]: This is something that a job launcher PR will have.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/RingTransmitFederateMain.cpp  2
         

Clones removed
==============
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

File: benchmarks/helics/EchoMessageHubFederateMain.cpp (hunk 8 above)
Comment (nightlark) [reply benchmarks/helics/EchoMessageHubFederateMain.cpp]: Yea, merging all of these into a single executable with subcommands is a good idea -- it would be a nice way to get rid of all the extra files.

File: benchmarks/helics/EchoLeafFederateMain.cpp (hunk 7 above)
Comment (phlptp) [reply benchmarks/helics/EchoLeafFederateMain.cpp]: If the file name contains the differentiation is might not be required but I suppose it would be nice to be able to get all the information in single block, instead of having to parse the file name and the file.  If it was contained in the file the processor could separate the steps and just parsing the file contents would have all the information.  So I think unless there is some issue with generating it, I think that information should be in the file contents as well.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/TimingHubFederateMain.cpp  2
- benchmarks/helics/EchoMessageHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitFederateMain.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  2
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  2
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  2
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/EchoLeafFederateMain.cpp  2
- benchmarks/helics/EchoHubFederateMain.cpp  2
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoLeafFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoMessageHubFederateMain.cpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/EchoHubFederateMain.cpp  1
- benchmarks/helics/TimingHubFederateMain.cpp  1
- benchmarks/helics/EchoLeafFederateMain.cpp  1
- benchmarks/helics/PholdFederateMain.cpp  1
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/RingTransmitMessageFederateMain.cpp  1
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/MessageExchangeFederateMain.cpp  1
- benchmarks/helics/EchoMessageLeafFederateMain.cpp  1
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederateMain.cpp  1
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/TimingLeafFederateMain.cpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/PholdFederate.hpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

File: benchmarks/helics/EchoHubFederate.hpp (hunk 5 above)
Comment (nightlark) [reply benchmarks/helics/EchoHubFederate.hpp]: I renamed it to `num_leafs`

File: benchmarks/helics/BenchmarkFederate.hpp (hunk 4 above)
Comment (nightlark) [reply benchmarks/helics/BenchmarkFederate.hpp]: The no arg constructor calls the one that creates the CLI11 app using an empty-string for the name.

File: benchmarks/helics/EchoLeafFederateMain.cpp (hunk 7 above)
Comment (nightlark) [reply benchmarks/helics/EchoLeafFederateMain.cpp]: I added a "benchmark federate type" output that shows the name of the federate app (EchoHub, EchoLeaf, etc).

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/ringBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by nightlark** [comment]
Travis CI builds need to get re-run tomorrow. Builds were failing before they got to our build scripts due to apt-get failures.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoLeafFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/ringBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

File: benchmarks/helics/BenchmarkFederate.hpp [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+  protected:
+    class Result {
+      public:
+        std::string name;
+        std::string key;
+        std::string value;
+    };
+    std::vector<Result> results; //!< a vector of output results to print
+
+    OutputFormat result_format{OutputFormat::PLAIN_TEXT}; //!< output format for printing results
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkFederate.hpp]: For packing purposes this variable could be put next to the bool variables at the end of the member list.  Doubt it matters much, but it is a benchmark.

File: benchmarks/helics/BenchmarkFederate.hpp [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+        // add common index options (optional)
+        opt_index = app->add_option("--index", index, "the index of this phold federate");
+        opt_max_index =
+            app->add_option("--max_index", maxIndex, "the maximum index given to a phold federate");
+        opt_max_index->ignore_underscore();
+
+        // add a flag for printing system info
+        app->add_flag_function(
+            "--print_systeminfo",
+            [](int count) {
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkFederate.hpp]: I think you could use `add_flag_callback` here since count isn't really doing much and --print_system_info=false is likely to still trigger it since that would result in a -1 in this case.

File: benchmarks/helics/BenchmarkMain.cpp [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            T().initialize("", "--help");
+            return std::string{};
+        });
+}
+
+int main(int argc, char* argv[])
+{
+    helics::helicsCLI11App app(
+        "HELICS benchmark federates for use in multinode benchmark setups", "helics_benchmarks");
+    app.ignore_case()->prefix_command();
```
Comment (phlptp) [inline benchmarks/helics/BenchmarkMain.cpp]: might want `->ignore_underscore()` here as well so you can do "echo_hub" as a subcommand and others as well.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

File: benchmarks/helics/BenchmarkFederate.hpp (hunk 11 above)
Comment (nightlark) [reply benchmarks/helics/BenchmarkFederate.hpp]: How will it help packing if OutputFormat isn't a bool?

Comment (phlptp) [reply benchmarks/helics/BenchmarkFederate.hpp]: most data types are 8 byte aligned on a 64 bit system.  So after result_format you have a 4 byte gap.  

right now after the bools you have a 6 byte gap,  so moving result_format would result in an overall size reduction of 8 bytes.  

Not critical but unless there is a good reason not to (like important stuff being in the first 64 bytes) I try to minimize the gaps and structure sizes.  In this case it may make sense to move the bools or leave them separate for readability.

Comment (nightlark) [reply benchmarks/helics/BenchmarkFederate.hpp]: Ah okay, so filling in the gap the bools are leaving open. I moved the bools up to below result_format, and rearranged the callbacks/fedptr/params -- if the size goes above 64bytes it would be nice to keep those ones together since could be accessed in the main loop.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
- benchmarks/helics/filterBenchmarks.cpp  2
         

Clones removed
==============
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/PholdFederate.hpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/echoBenchmarks.cpp  -13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

File: benchmarks/helics/TimingLeafFederate.hpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    void doFedInit() override
+    {
+        pub = fed->registerPublicationIndexed<std::string>("leafsend", index);
+        sub = fed->registerSubscriptionIndexed("leafrx", index);
+    }
+
+    void doMainLoop() override
+    {
+        int cnt = 0;
+        // this is  to make a fixed size string that is different for each federate but has sufficient length to
```
Comment (phlptp) [inline benchmarks/helics/TimingLeafFederate.hpp]: this comment can be removed

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- benchmarks/helics/echoMessageBenchmarks.cpp  3
- benchmarks/helics/EchoLeafFederate.hpp  1
- benchmarks/helics/RingTransmitFederate.hpp  1
- benchmarks/helics/RingTransmitMessageFederate.hpp  1
- benchmarks/helics/filterBenchmarks.cpp  2
- benchmarks/helics/EchoMessageLeafFederate.hpp  1
         

Clones removed
==============
+ benchmarks/helics/echoBenchmarks.cpp  -13
+ benchmarks/helics/ringMessageBenchmarks.cpp  -1
+ benchmarks/helics/timingBenchmarks.cpp  -16
+ benchmarks/helics/ringBenchmarks.cpp  -1
+ benchmarks/helics/PholdFederate.hpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5048917&bid=16624558)

**Comment by nightlark** [comment]
Yea, I saved runs of the single machine benchmark results before making changes and the times after are about the same (normal variation, sometimes runs are faster/slower).

**Comment by phlptp** [comment]
IS this ready to merge?
